Supertrait Auto-impl

dingxiangfei2009 opened the pull request:
We would like to propose `auto impl` syntax for supertraits for a few language enhancement, to ultimately facilitate easier trait evolution and refactoring and easier trait authoring when trait hierarchy is in consideration.

24/09/2025: revision published with typo fixes and dropping the `unsafe` qualifier for `probe_extern_impl` attribute.

<details><summary>Past edit history</summary>
21/09/2025: revision published with the following changes.
- Address [comments from @cramertj](https://github.com/rust-lang/rfcs/pull/3851#discussion_r2301552062) by including the worked examples mentioned in the [design meeting](https://hackmd.io/@1n2nRkvSTd-QgK8cCPds1w/HkBmeCE7xx). A new proposal is included so that an unsafe attribute may be used to allow compiler to probe further and decide if `auto impl` default implementation would be conflicting. This might need consultation with the types team as its effectiveness and stability depends very much on the capabilities of the trait solver.
- Address [comments from @ElchananHaas](https://github.com/rust-lang/rfcs/pull/3851#discussion_r2306004253) and [comments from @N4tus](https://github.com/rust-lang/rfcs/pull/3851#issuecomment-3239334843) on several deficiencies in the worked examples, such as missing superbounds and complexity in the motivating examples.
- Address [a comment from @programmerjake](https://github.com/rust-lang/rfcs/pull/3851#discussion_r2302020617) by removing the template text.
</details>

[Rendered](https://github.com/dingxiangfei2009/rust-rfcs/blob/supertrait-auto-impls/text/3851-supertrait-auto-impls.md)

dingxiangfei2009 committed:
supertrait auto impl RFC v2

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

ElchananHaas commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,642 @@
+
+For some traits, it's difficult to implement the trait directly because the "raw" interface that the trait exposes is complex. If the trait is unsafe, this may be even worse as the end-user may not wish to write any unsafe code. This feature makes it easy to provide utilities for more easily implementing the trait.
+
+### Example: Serde
```
It would be better to use an example of an existing trait and trait evolution since that is the primary purpose of Supertrait auto impl.

dingxiangfei2009 replied:
As requested, I am also adding more worked examples from trait evolution.

N4tus commented:
The explanation states that the traits need to have a super/sub trait relation. But some examples use traits e.g. `MouseEventHandler` and `Double`, that are not in a relation to the trait they provide an `auto impl` for.

dingxiangfei2009 committed:
apply suggestions and propose extensions

dingxiangfei2009 commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,642 @@
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
```
Understood. This was from the template text. It should have been dropped.

dingxiangfei2009 commented on text/0000-supertrait-auto-impls.md:
````diff
@@ -0,0 +1,642 @@
+    }
+}
+```
+And then `MyStruct` automatically implements `Serialize` by creating a `MyStructProxy` instance and serializing the proxy. So for example `MyStruct { name: "a", int: 42 }` is serialized into json as `{"name":"a","tens":4,"digit":2}`.
````
I am adding additional explanation.

dingxiangfei2009 commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,642 @@
+    type Type1;
+    type Type2;
+}
+trait Subtrait: Supertrait1 + Supertrait2 {
```
I am simplifying the example a bit.

programmerjake commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -55,15 +221,36 @@ struct MyStructProxy<'a> {
+    where Self: 'a;
+
     fn serialize(&self) -> MyStructProxy<'_> {
         Proxy {
```
```suggestion
        MyStructProxy {
```

programmerjake commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -498,6 +734,100 @@ If the sub-trait defines an item of the same name as an item in the super-trait,
+| Supertrait kind | `auto impl` block in sub-`trait` block | `auto impl` statement in sub-`trait` block |
+|-|-|-|
+| non-marker | Mandatory[<sup>a</sup>](#e-i-a) | Optional[<sup>b</sup>](#e-i-b) |
+| marker <td colspan="2" style="text-align: center"> Mandatory[<sup>c</sup>](#e-i-c) |
```
this doesn't render correctly.
```suggestion
| marker | Mandatory[<sup>c</sup>](#e-i-c) | Mandatory[<sup>c</sup>](#e-i-c) |
```

programmerjake commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -498,6 +734,100 @@ If the sub-trait defines an item of the same name as an item in the super-trait,
+The reason for this is that given that `trait Subtrait` has already provided its implementation, an implementation of `Subtrait` must choose between the default implementation and a user-defined implementation. We prefer explicit confirmation through `extern impl` declaration from the implementor, rather than making the compiler to reason about whether `auto impl` should be backfilled for ease of ...
+
+#### Extension: Possible relaxation through an unsafe attribute and a future-compatibility lint
+For important ecosystem traits like `PartialOrd` and `Ord`, this rule is still unsatisfactory due to [the potential rewrites required](#po-o) on downstream crates, even though it could be as small as an additional `extern impl PartialOrd`. As an extension, the rule could be relaxed with an unsafe attribute `#![unsafe(probe_extern_impl)]` and apply further trait selection to decide whether the def ...
```
what operation is this performing that could produce UB therefore requiring `unsafe`? if there is none, `unsafe` should not be used.

dingxiangfei2009 commented:
Quick question for @fmease, should I rename the RFC file from serial `0000-...` to `3851-...` in this text by ourselves? Or will there be a bot to rename the file?

dingxiangfei2009 committed:
do not unsafe the probe_extern_impl attribute; fix rendering and typos

dingxiangfei2009 committed:
rename as per RFC number

dingxiangfei2009 commented on text/0000-supertrait-auto-impls.md:
```diff
@@ -577,14 +907,22 @@ trait AutoMyTrait: MyTrait {
-impl MyOtherTrait for Foo {} // <-- now it should not be generated anymore
+impl MyOtherTrait for Foo {}
+
+// QUESTION: which `impl MyTrait for Foo` fulfills the boud `Foo: MyTrait`?
```
Applied

programmerjake commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,975 @@
+The reason for this is that given that `trait Subtrait` has already provided its implementation, an implementation of `Subtrait` must choose between the default implementation and a user-defined implementation. We prefer explicit confirmation through `extern impl` declaration from the implementor, rather than making the compiler to reason about whether `auto impl` should be backfilled for ease of ...
+
+#### Extension: Possible relaxation through an unsafe attribute and a future-compatibility lint
+For important ecosystem traits like `PartialOrd` and `Ord`, this rule is still unsatisfactory due to [the potential rewrites required](#po-o) on downstream crates, even though it could be as small as an additional `extern impl PartialOrd`. As an extension, the rule could be relaxed with an unsafe attribute `#[probe_extern_impl]` and apply further trait selection to decide whether the default impl ...
```
and here

programmerjake commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,975 @@
+
+In this scenario, any item in an impl block of the sub-trait using the ambiguous name will always refer to the item from the sub-trait. This means that the only way to override the item from the super trait is to use `extern impl`.
+
+## Mandatory `external impl` declaration
```
```suggestion
## Mandatory `extern impl` declaration
```

programmerjake commented on text/3851-supertrait-auto-impls.md:
````diff
@@ -0,0 +1,975 @@
+It's possible to declare that an auto implementation is unsafe.
+```rs
+trait MySubTrait: MyTrait {
+    unsafe auto impl MyTrait;
````
so, does this create an `unsafe` requirement or discharge an existing `unsafe` requirement? you seem to have it do both which is generally not how Rust uses `unsafe` anymore, now that we have [`unsafe_op_in_unsafe_fn`](https://doc.rust-lang.org/edition-guide/rust-2024/unsafe-op-in-unsafe-fn.html) enabled by default

dingxiangfei2009 replied:
I see that this can be confusing.

- Inside a `trait Subtrait: .. { .. }` block, `unsafe auto impl` creates an `unsafe` requirement.
- Inside a `impl Subtrait { .. }` block, `unsafe auto impl` discharges the corresponding `unsafe` requirement.

Should we adjust the "keyword" orders, so that the difference can be more obvious?

cramertj replied:
+1 to @programmerjake 's expectations-- I would've expected `unsafe auto impl` to discharge an unsafe requirement-- that is, guarante that the provided impl of the supertrait fulfills the requirements of the `unsafe trait SuperTrait`. This is the usual meaning of `unsafe impl` in Rust.

dingxiangfei2009 committed:
remove unsafe attribute; fix typo

cramertj commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,975 @@
+
+It is still up for discussion.
+
+## Why use `unsafe` in this way
```
The RFC no longer uses `unsafe` in a surprising way, so I think this section can be dropped.

programmerjake replied:
I think the RFC still uses `unsafe` in a surprising way: see [this thread](https://github.com/rust-lang/rfcs/pull/3851#discussion_r2376681991)

obi1kenobi commented:
This is a very interesting proposal!

As maintainer of `cargo-semver-checks`, I'm thinking about the SemVer implications here. Due to the significant number of edge cases (associated types, associated functions, `auto impl`, `extern impl`, etc.), I'm quite sure I would fail if I attempted to enumerate all the possible SemVer hazards myself. At the same time, I think it's very important that SemVer hazards be considered explicitly — especially for a complex language feature like this.

Would you mind adding a `## SemVer hazards` section to the doc and enumerating all the new ways to cause a major breaking change introduced here?

dingxiangfei2009 commented:
@obi1kenobi Yes, I remember that. I will tag you in the "changelog" when I add the SemVer hazard section for your reviews.

dingxiangfei2009 committed:
additional motivation with Cow; semver consideration

dingxiangfei2009 committed:
November edition

- Clarify the rule on possible nesting
- Clarify the semver hazards
- Clarify the name resolution ambiguity from two supertraits

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

dingxiangfei2009 commented:
November update
- I added a rule to clarify that nesting `auto impl`s could happen and how the nesting would work.
- I added a section on SemVer and explored the consequence from possible edits to traits using this feature. cc @obi1kenobi 
- I clarified the name resolution ambiguity when the name can be resolved to two supertrait items.

dingxiangfei2009 committed:
Mention the tower::Service trait for motivation

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

dingxiangfei2009 committed:
extension: lifetime generic auto impl

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

lcnr commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,975 @@
+// and we deduce based on some applicability criterion ...
+impl AutoMyTrait for Foo {} // <-- generates MyTrait
+
+// Suppose this item is added at one point of time
```
This came out of a meeting with @joshtriplett and @tmandry. Implementing "do these impls overlap" via proper coherence checking is significantly harder than doing some self-type based fast reject.

Right now builtin auto-trait impls are disabled if there exists any impl with the outermost rigid constructor (e.g we disable all auto trait impls for `MyType<T>` if there's an impl for `MyType<i32>` cc https://github.com/rust-lang/rust/issues/93367). Doing something similar here is kinda trivial from a type system pov

The more powerful "whether we emit an impl depends on whether coherence thinks impls overlap" is a lot harder as suddenly the while trait system needs to reason about "impls may or may not exist at this point". Supporting that requires a new `TypingMode` and refactorings to the way we iterate trait impls in the solver and so on. That has a significantly higher type system cost.

I personally would kinda prefer not doing proper trait solving:tm: to figure out whether to emit impls, but hmm. I guess I've become more open to make the type system harder more involved over time it's nice because the typing mode used to figure out whether to add these impls won't be soundness critical, so it's just annoying to impl

dingxiangfei2009 committed:
apply suggestions; add one more refactor example

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
````diff
@@ -0,0 +1,1208 @@
+}
+```
+
+There are now two choices for type `X` in the downstream crate.
````
Is this a *forced* choice, meaning that `X` downstream is broken unless it chooses one of these (new) choices?

AFAICT this is a major breaking change for the stdlib, because `extern impl PartialOrd;` appears to be required to support separate `impl PartialOrd for X` and `impl Ord for X`. If that's the case, it might be good to mention that how this would be addressed will be explained later, ideally in a named section that can have its anchor linked here.

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+
+### Example: Serde
+
+The serde traits are notoriously difficult to implement directly. It's almost always done by macro. Imagine if you could write this:
```
I had to look up the `serde::Serialize` trait to make sense I understood this example correctly:
```
pub trait Serialize {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
       where S: Serializer;
}
```

It might be useful to mention it here and save the reader the lookup.

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+  - We need to maintain that all the current `impl` blocks to compile, specifically the `impl Deref` litmus test mentioned in the [deref-receiver] motivating example.
+  - This demand also extends to other possible library traits that may see relocation of items into a future supertrait, while ensuring that the existing `impl` blocks continue to compile.
+- Intuitional readability and clear syntatical signal
+  - We strive for a syntax that is intuitional and easily connected to the existing constructs.
```
typo, I believe:
```suggestion
- Intuitive readability and clear syntatical signal
  - We strive for a syntax that is intuitive and easily connected to the existing constructs.
```

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+All items inside the `auto impl` block must match an item from the super trait of the same name and signature. When the block is empty, it is legal to use a semicolon instead. That is, these are equivalent:
+
+- `auto impl SuperTrait {}`
+- `auto impl SuperTrait;`
```
I assume this means that if `SuperTrait` has no items, either of these forms counts as a complete implementation of `SuperTrait` as well when the subtrait that includes this `auto` block is implemented, correct?

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+
+## Unsafe auto implementations
+
+The `auto impl` items can be marked `unsafe`, which declares that implementing the sub-trait without using the auto implementation is unsafe.
```
Does this need to be updated per the use-of-`unsafe` discussion above? This seems to introduce a safety obligation, but `unsafe auto impl` in comments above was described as discharging obligations, not introducing them.

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+
+This is a SemVer hazard and mandates a major change. The justification follows API change in trait irregardless of super- or sub-trait relationship. This scenario encompasses any changes in types, function signature, bounds, names.
+
+---
```
One more example:

Adding defaulted items to a trait is usually non-breaking:
```rust
pub trait Super {
    // Adding this item is fine.
    fn example() -> i64 { 0 }
}
```

But say a subtrait contains two `auto impl` supertraits:
```rust
pub trait Subtrait {
    auto impl upstream::Super;
    auto impl other::Other;
}
```

What happens if `other::Other` already had an `example` function?

If I've understood the "Naming ambiguity" portion of the proposal correctly, this is now an ambiguity conflict:
> If the sub-trait definition contains two `auto impl` directives and a sub-trait implementation has an item with a name that can be resolved to an associated item in both of the `auto impl` supertraits, irrespective of the associated item kind, then it **must** also be rejected as ambiguity.

So now adding a trait item with a default is newly SemVer-major as a result of this feature, while previously it was totally safe! Oops!

dingxiangfei2009 replied:
Maybe the wording is not perfect, so let me attach a worked example in https://github.com/rust-lang/rfcs/pull/3851/commits/25597e7b584d355c4fb6fe170e26ee734e3d42ff.

So, if I run through the example, no matter how many items with default implementation (actually I think we should call them provided methods?) are added into `TraitA` or `TraitB`, they are not considered for name resolution- one just cannot provide another implementation in their own trait definition nowadays.

So the wording should have been, name ambiguity only arises when in `impl Subtrait` there is an item `foo` that matches a **required** item `trait TraitA` and another in `trait TraitB`, while in `trait Subtrait` definition `foo` is missing.

```rust
pub trait TraitA {
    fn foo();
}
pub trait TraitB {
    fn foo();
}
pub trait Subtrait: TraitA + TraitB {
    auto impl TraitA;
    auto impl TraitB;
}
impl Subtrait for MyType {
    fn foo() {} //~ ERROR This is true name ambiguity.
}
```

And under no circumstances `Subtrait::foo` can be resolved given the above example. I do not intend to make any `$name` in `Subtrait::$name` to be resolved to any item not truly belonging to `Subtrait`. This is one invariant I do not wish to break.

obi1kenobi reviewed (commented):
Apologies for not being able to take another look at this sooner. I went through the RFC from start to finish, and left my thoughts inline.

Unfortunately, I think the SemVer hazards section needs *a lot* more work.

Among other concerns, if I understand the semantics correctly, I believe this proposal takes several previously SemVer-minor changes and upgrades them to SemVer-major in an unexpected way. The implications of that are quite serious! We need to figure out if that's a deal-breaker, something that can be avoided with more design work, or an acceptable and regrettable casualty of shipping this desirable feature. In any case, I think the RFC needs to tackle that head-on.

dingxiangfei2009 commented on text/3851-supertrait-auto-impls.md:
````diff
@@ -0,0 +1,1208 @@
+    fn click_event(&mut self, event: ClickEvent);
+    fn move_event(&mut self, event: MoveEvent);
+}
+```
````
Thank you for your comments! Let me open my comment with response to this concern.

>I think this is a fairly serious flaw for a feature that aims to _improve_ the overall situation of what’s possible in semver-comparible API changes.

So I do not believe that improving the semver-compatibility is a goal of this proposal and I do not try to sell this proposal as a general solution to the problem.

One of the motivations of the proposal is to provide a more user friendly and flexible alternative to blanket implementations involving supertraits without the problems coming from them such as potential overlapping implementations and inability to override those blanket implementations. As illustration, let us use the `EventHandler` and see what `auto impl` offers and _does not offer._

Without `auto impl`, the way to supply default implementation is to introduce a blanket implementation.

```rust
// CRATE A
trait MouseEventHandler: EventHandler {
    fn click_event(&mut self, event: ClickEvent);
    fn move_event(&mut self, event: MoveEvent);
}

impl<T: MouseEventHandler> EventHandler for T { // (*)
	type Event = MouseEvent;
	fn handle_event(&mut self, event: MouseEvent) {
		use MouseEvent::*;
		match event {
			ClickEvent(evt) => self.click_event(evt),
			MoveEvent(evt) => self.move_event(evt),
		}
	}
}

// CRATE B
trait NicerMouseEventHandler: MouseEventHandler { .. }

impl<T: MouseEventHandler> NicerMouseEventHandler for T { .. }
```

Now with `auto impl` as described in the RFC, for each `impl MouseEventHandler for MyType` for some `MyType` there will be an `impl EventHandler for MyType` that is templated after the `auto impl` block. The blanket implementation, as we know that comes with quite some drawbacks, is not necessary.

What I advocate strongly is that `auto impl` should still be treated conceptually as a blanket implementation with some modification. To think about it, we should ask ourselves, when the blanket implementation `(*)` in Crate `A` changes, by means of changes to the associated type **or changes to implementation details of `handle_event` even**, do we introduce SemVer breaking change? Please correct me if I am wrong, I think _it is definitely a SemVer breaking change hazard._ It could be overarching to say that any implementations with generics, blanket implementations included, will induce SemVer hazards when they are changed even in implementation details, but it rule sounds reasonable to me. cc @obi1kenobi

And I don't think this view should be altered when `auto impl` is morally a blanket implementation, so questions about SemVer hazards should be resolved by treating them as blanket implementations and answered in that domain.

> If you imagine such implementations helpers may even come from a different crate than EventHandler itself, and yet-another crate ultimately uses it downstream for defining PrintHandler, then any change switching out the convenience/helper used to power the EventHandler impl would in and my itself also be a breaking change.

So, yes, following this train of thought these are all breaking changes.

> In my opinion, ideally, it should somehow be made possible to implement something _like_ a trait, but that isn’t necessarily an _actual_ trait itself, but merely functions like a template for implementing one (or several) other traits more conveniently.

> It’s unfortunate that this seems unnecessary for the _primary_ motivating use-case of updating true hierarchies where all traits involved and all methods _are_ supposed to be public and stably available anyway, not mere implementation details of a kind of trait-to-trait adapter; but I’m 100% certain that a supertrait auto-impl like this one _will_ be used for that kind of thing, no matter what, so a more complete solution design seems somewhat necessary IMHO.

I still do not quite get what you are getting to and I am curious about your ideas. Would you mind elaborating? A few sketches will help.

obi1kenobi replied:
> So I do not believe that improving the semver-compatibility is a goal of this proposal and I do not try to sell this proposal as a general solution to the problem.

This raises the obvious question: should it be a goal?

I believe we should explore what it would take to get to "yes".

Blanket impls are full of surprising SemVer hazards today. This proposal in its current state would make things even more challenging. It would be the wrong tradeoff, in my opinion, to address one problem by introducing a harder one.

dingxiangfei2009 replied:
Replying to @steffahn ...

#### On `template trait`

I like the `template trait` idea honestly. I have one question. In the `BorrowOrdHash` we have apparently the item `borrow` syntatically associated to it. `BorrowOrdHash` is not a trait, so `borrow` is an inherent method to the `Self` type? What is the implication to the method resolution?

On the overall topic, I wonder this can be solved by existing unstable features like rust-lang/rust#41517.

It is also a matter of perspective. To me `trait` is a contract and a set of invariants. Following this strict interpretation, which I recommend every user to follow, no two traits are SemVer compatible as soon as there is a mismatch in the crate identity or version, no matter how void the trait content would be. Moving from `EqOrdByKey + HashByKey` to  `BorrowOrdHash` is a breaking change and I will not have a second thought when I bump my crate version.

However, this is a common pattern in ecosystem to have a bag of trait bounds and give it a name. What they want is have a way to write a trait bound succinctly instead of writing 42 trait bounds connect with `+` in their `where` clauses everywhere. For that we already have answer, which is trait alias. If one really wants a bag of traits, we should make it work.

On that note, `template trait` can fill a gap between trait alias + bag of impls and `auto impl`, on the point of introducing SemVer hazard. I would say SemVer hazard is not necessarily a bad thing, it is _good_ by a lot. It is rather a bad idea that any type that implements this bag of traits `Eq`/`Ord`/... could be used to call my API. That would mean that I loose my control over what types are safe to use for my library, because the user could call my library without generating the right impls through (proc-)macros or writing the right impls because they never read trait documentation.

So in my opinion when a trait is really so void of trait characteristics and regresses to just a bag of trait impls without strings attached, I would recommend to make trait alias work, which I am interested in also because of its important niche, and let our users use it. To throw right tools at the problem basically.

---

#### On orphan rules

`auto impl` should not allow any changes to trait solving or trait selection at this point, this is agreed upon. On the point of orphan rules, `auto impl` does not intend to make `impl<T> Foo<MyStruct, T> for String {}` compile-what a blanket implementation cannot do, should `auto impl` will not enable either. If we really need to be careful, I think this is about good diagnostic to explain why orphan rule is violated at which `impl`.

dingxiangfei2009 committed:
Resolve ambiguous wording on name collision with supertraits

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+This is a SemVer hazard and mandates a major change. The justification follows API change in trait irregardless of super- or sub-trait relationship. This scenario encompasses any changes in types, function signature, bounds, names.
+
+---
+
```
I reject this line of argument in the strongest terms.

Quite honestly, I find your reply insulting. Especially so given the amount of time and energy I've already invested in both cargo-semver-checks and in this RFC.

I've demonstrated in other comments that this is absolutely not "as if" equivalent to blanket implementations. It is disingenuous to claim this is trivial to work out just by starting with the rules for blanket implementations.

Even if it were, I feel it is not acceptable for a proposer of a new language feature to demand someone else work out its edge cases.

dingxiangfei2009 replied:
It is unfortunate that our conversation arrived at this state. I would like to clarify that I do not intend to downplay the importance of `cargo-semver-checks` and I have greatly and respectfully reviewed your comments insofar. 

However, I have a hard time to understand exactly how `cargo-semver-checks` would consider, whether SemVer hazard would take place under which rules. I have attended to your talks and I remember that there have been many many rules built into this tool. I would like to seek your understanding that without a clear deduction formulation of SemVer hazards, language feature designers could find themselves paralysed and confused. Please understand that, given that we are at an early stage of exploration, I would prefer the RFC text to stay focused on the main language design. I am at the same time sympathetic towards the ecosystem tool designer and, therefore, I offered an option to simplify the the SemVer deduction by reducing it and modelling it after blanket implementation analogy. That is my suggestion and I am eager to hear if this model fails at any other edge cases.

For the time being, I am sad to say that I have not enough expertise to expand the SemVer hazard section to the extend that meets the rigor that `cargo-semver-checks` requires, not until I become the expert in the tool itself. Until then, I would still need your help and I have always appreciated it. 🙇

obi1kenobi commented on text/3851-supertrait-auto-impls.md:
```diff
@@ -0,0 +1,1208 @@
+
+This is a SemVer hazard and mandates a major change. The justification follows API change in trait irregardless of super- or sub-trait relationship. This scenario encompasses any changes in types, function signature, bounds, names.
+
+---
```
> > // in crate `upstream2`, which depends on `upstream1`
> 
> Also this will not compile. `Supertrait` has an item `Item` but your `auto impl Supertrait` block is empty.

Could you point me to where the RFC specified this requirement? I must have missed it.

This is why your help is key in figuring out the edge cases. This is a complex feature, and you understand it 10x better than anyone else.

tmandry replied:
> Just to clarify, do you mean "Adding an auto impl item to an existing trait is only a non-breaking change if the auto impl is for a trait within the same crate"? If so, that seems fine.

Basically, yes. In fact I was talking about something slightly more narrow than this: That adding a defaulted item to a new or existing auto impl, when the item being defaulted is defined in another crate, is a breaking change.

---

This is different from the case where you say "my existing item `foo()` now lives in this supertrait, your impls of my `foo()` now implement `Supertrait::foo()`". For that see my suggestion in https://github.com/rust-lang/rfcs/pull/3851#discussion_r2933383431.

Thinking about the two combined, I think we would we end up with these implications:

1. Replacing an existing item with a reference to a supertrait item in another crate is fine if you explicitly name it from the subtrait (`use Supertrait::item;` or similar, as in the linked comment).
2. It is ok to default that item if it was already defaulted, _or_ at the same time as replacing it.
3. Once you release a version moving the item to the supertrait without defaulting, you lose the ability to add a default later.

edit: My line of reasoning is that following these implications offloads the problem of "which default is this item using" to the existing problem of "are you using this auto impl or not".

dingxiangfei2009 commented on text/3851-supertrait-auto-impls.md:
````diff
@@ -0,0 +1,1208 @@
+        println!("my_second_func on String");
+    }
+}
+```
````
Yes, it is breaking change. In fact, I see merits in the sense that `extern impl Supertrait` would now be required in `impl Subtrait for S {}` block, as confirmation that
- the criterion to implement `Supertrait` is met, which is very important if `Supertrait` is a marker trait;
- the choice between `auto impl` in `trait Supertrait` or `extern impl` while keeping the existing `impl Supertrait` is explicit.

For this reason, I think it is desirable and I do not quite agree with it being footgun-y.

dingxiangfei2009 committed:
Clarify the name ambiguity rules by providing a worked example

dingxiangfei2009 committed:
Strengthen then ambiguity resolving wording

In name resolution, we should only care about required items.

dingxiangfei2009 committed:
Refine wording around use of `unsafe`

Signed-off-by: Xiangfei Ding <dingxiangfei2009@protonmail.ch>

lygstate commented:
Suggest avoid the auto keyword. use existing impl and extern keyword is enough.

```rust
trait Supertrait1 {
    type Type;
}
trait Supertrait2 {
    type Type1;
    type Type2;
}
trait Subtrait1: impl Supertrait1 {
}
impl Subtrait1 for MyType {
    type Type = u8; // implicitly implements `Supertrait1::Type := u8`
}

trait Subtrait2: impl Supertrait2 {
    impl Supertrait2 {
        type Type1 = Self;
        type Type2 = ();
    }
}
impl Subtrait2 for MyType {
    // An implicit `impl Supertrait2 for MyType` is generated
    // with `Type1 := MyType` and `Type2 := ()` as backfill
}
```
duplicated code reduced. 


```rust
pub trait TraitA {
    fn foo();
}

pub trait TraitB {
    fn foo();
}

pub trait TraitC {
    fn foo();
}

pub trait Subtrait: impl TraitA + impl TraitB + TraitC {
    fn foo();         // Also OK (*)
}

// Now, `Subtrait::foo` always resolves to the definition (*)

impl Subtrait for MyType {
    impl TraitA {
        fn foo() {}
    }
    impl TraitB {
        fn foo() {}
    }
    fn foo() {} // (**)
}

impl TraitC for MyType {
    fn foo() {}
}

let _ = <MyType as Subtrait>::foo(); // OK, `Subtrait::foo` resolves to the instance (**)
// NOT `<MyType as TraitA>::foo` or `<MyType as TraitB>::foo`
```

[Rendered](https://github.com/lygstate/rust-rfcs/blob/supertrait-auto-impls/text/3851-supertrait-auto-impls.md
)
